src/precompiles/keccak256.rs

```diff
@@ -0,0 +1,194 @@
+use super::{precompile_abi_in_log, MemoryLocation, MemoryQuery, Precompile};
+use crate::{eravm_error::EraVmError, heaps::Heaps};
+use crypto_common::hazmat::SerializableState;
+use sha3::{Digest, Keccak256};
+use u256::U256;
+
+pub const KECCAK_RATE_BYTES: usize = 136;
+pub const MEMORY_READS_PER_CYCLE: usize = 6;
+pub const KECCAK_PRECOMPILE_BUFFER_SIZE: usize = MEMORY_READS_PER_CYCLE * 32;
+pub const MEMORY_WRITES_PER_CYCLE: usize = 1;
+
+pub struct ByteBuffer<const BUFFER_SIZE: usize> {
+    pub bytes: [u8; BUFFER_SIZE],
+    pub filled: usize,
+}
+
+impl<const BUFFER_SIZE: usize> ByteBuffer<BUFFER_SIZE> {
+    pub fn can_fill_bytes(&self, num_bytes: usize) -> bool {
+        self.filled + num_bytes <= BUFFER_SIZE
+    }
+
+    pub fn fill_with_bytes<const N: usize>(
+        &mut self,
+        input: &[u8; N],
+        offset: usize,
+        meaningful_bytes: usize,
+    ) {
+        assert!(self.filled + meaningful_bytes <= BUFFER_SIZE);
+        self.bytes[self.filled..(self.filled + meaningful_bytes)]
+            .copy_from_slice(&input[offset..(offset + meaningful_bytes)]);
+        self.filled += meaningful_bytes;
+    }
+
+    pub fn consume<const N: usize>(&mut self) -> [u8; N] {
+        assert!(N <= BUFFER_SIZE);
+        let mut result = [0u8; N];
+        result.copy_from_slice(&self.bytes[..N]);
+        if self.filled < N {
+            self.filled = 0;
+        } else {
+            self.filled -= N;
+        }
+        let mut new_bytes = [0u8; BUFFER_SIZE];
+        new_bytes[..(BUFFER_SIZE - N)].copy_from_slice(&self.bytes[N..]);
+        self.bytes = new_bytes;
+
+        result
+    }
+}
+
+fn get_state_bytes(bytes: &[u8]) -> [u64; 25] {
+    let mut result = [0u64; 25];
+
+    // grab in chunks of 4 bytes and reverse them
+    for i in 0..25 {
+        let mut chunk = [0u8; 8];
+        chunk.copy_from_slice(&bytes[(i * 8)..((i + 1) * 8)]);
+        result[i] = u64::from_le_bytes(chunk);
+    }
```

Comment:
The comment contradicts the code here. It's 8 bytes chunks, not 4 bytes. Which one is correct?